rfcs/0000-commonmark-docs.md

```diff
@@ -0,0 +1,304 @@
+- Facebook's [React documentation][react-docs] (using Gatsby with
+  plain CommonMark)
+- Microsoft's [VS Code documentation][vscode-docs] (using Jekyll)
+- [Kubernetes documentation][kubernetes-docs] (using Hugo)
+- The [Rust book][rust-docs] (using mdBook)
+
+The goal of this RFC is to change the *form* of the current
+documentation for Nix, Nixpkgs and NixOS to look similar to any one of
+the above 5 projects (see requirements in the next section). We submit
+that the least effort route to do so is to use the same toolchain as
+they do (CommonMark or MDX input and one of Gatsby, Jekyll or Hugo to
+generate a static site).
+
+What all of the projects cited above have in common is a large
+contributor base with heterogeneous skillsets and multiple
+subcommunities. Just like in Nixpkgs, they cannot leverage the
+hegemony of RST in Python, because not all their users are Python
+programmers. They prefer not to count on writers learning Docbook or
+Asciidoc, because many documentation patches are from casual
+contributors. The *lingua franca* across subcommunities for both
+humans and their toolchains is Markdown. It's all the more easy to
+create good looking documentation with Markdown that the tools
+available to process it are plentiful and flexible (JavaScript
+converters with plugin support like [Remark][remark], [static site
+generators][staticgen] by the dozen, [MDX][mdx] to extend CommonMark with
+arbitrary React components, etc).
+
+[gatsby-docs]: https://www.gatsbyjs.org/docs/
+[gitstar-rankings]: https://gitstar-ranking.com/repositories
+[kubernetes-docs]: https://kubernetes.io/docs/home/
+[react-docs]: https://reactjs.org/docs/getting-started.html
+[rust-docs]: https://doc.rust-lang.org/book/
+[tensorflow-docs]: https://github.com/tensorflow/docs
+[vscode-docs]: https://code.visualstudio.com/docs
+[mdx]: https://github.com/mdx-js/mdx
+[remark]: https://remark.js.org/
+[staticgen]: https://www.staticgen.com/
+
+# Detailed design
+[design]: #detailed-design
+
+AsciiDoc, CommonMark and RST are all formats that support basic
+markup: emphasis, bold, (nested) (un)numbered lists, headings, inline
+and display code, tables (as HTML in the case of CommonMark), etc. The
+requirements we set below pertain to the appearance of the
+documentation available on the website.
+
+The key requirements we work towards are:
+
+1. easy-to-find "Edit on GitHub" button to increase drive-by
+   contributions,
+1. good quality documentation search engine,
+1. syntax highlighting of all code,
+1. separate page per chapter, instead of the current monolithic page
+   for each manual,
+1. table of content for each chapter available in a side bar to easily
+   jump through long content,
+1. make warning and info boxes stand out from the rest of the content
+   with colour,
+1. easy to customize HTML and CSS, since tired old templates we've all
+   seen too many times are best avoided. The objective is to make
```

Comment:
Here's my opinion about that:

First, it's a branding thing, *mdbook* AFAIK was made for the Rust documentation, so it's already their branding for their docs. 

Then, this simple objective is a way to avoid *locking ourselves* into what the tool provides. It may be that we need such capabilities not to brand the docs, but to implement missing features according to our needs. If extensibility is hard, or impossible, it's like painting us in a corner.

Finally, a throwback to the first point, is that we don't want to *lose* features. The current documentation toolchain allows a tight integration of the docs into the website. If the chosen solution doesn't allow us to integrate the docs in the website "as native as it can be", it's a loss of functionality.

Reply:
`mdbook` is customizable, see https://rust-lang.github.io/mdBook/format/theme/index.html. However, I don't think the current tight integration (e.g. that the docs have the same navbar as the rest of the site) is essential. I think it would be enough to have a visual style that looks similar to the rest of the site.

Or we could just upload the docs to readthedocs.io...

Reply:
> _mdbook_ AFAIK was made for the Rust documentation, so it's already their branding for their docs.

Ah, I didn't know that.

Reply:
We should have a `highlight.js` that can provide syntax highlighting for nix and bash.

Reply:
Started a mdBook docset for a different project lately, and found it's extremely easy (like most things with mdBook so far) to add additional languages for highlight.js.

I downloaded a script with the languages I needed from https://highlightjs.org/download/ and dropped it in `src/highlight.js`.
The available syntax highlighting for Nix also works great, of course.

One thing that irritates me is the default code-block width, but we could style this as needed.

Without said custom styles, I had to add a few line breaks to large expressions, such as <a href="https://github.com/Infinisil/system/blob/eb1046aa55802989ac8baffc8f476a5e3a682451/config/new-modules/default.nix">this one</a>.
Otherwise, it only displayed ~50% of the code block.

<details>
<summary>Small code block width</summary>

![image](https://user-images.githubusercontent.com/1847524/88964093-d8977580-d276-11ea-8e47-84778293bfc4.png)
</details>

Reply:
(In case it wasn't obvious, I wasn't talking badly about mdbook, I haven't looked into any of the tooling options since anyway this RFC is not about tooling.)

Reply:
@evanjs Thanks, I've added a `highlight.js` with Nix support to the manual. Example: https://hydra.nixos.org/build/124655688/download/1/manual/expressions/language-constructs.html